core/src/main/java/com/linecorp/armeria/common/stream/ThreeElementFixedStreamMessage.java

```diff
@@ -0,0 +1,198 @@
+    final void cleanupObjects(@Nullable Throwable cause) {
+        if (obj1 != null) {
+            try {
+                onRemoval(obj1);
+            } finally {
+                StreamMessageUtil.closeOrAbort(obj1, cause);
+            }
+            obj1 = null;
+        }
+        if (obj2 != null) {
+            try {
+                onRemoval(obj2);
+            } finally {
+                StreamMessageUtil.closeOrAbort(obj2, cause);
+            }
+            obj2 = null;
+        }
+        if (obj3 != null) {
+            try {
+                onRemoval(obj3);
+            } finally {
+                StreamMessageUtil.closeOrAbort(obj3, cause);
+            }
+            obj3 = null;
+        }
+    }
+
+    @Override
+    public void request(long n) {
+        final EventExecutor executor = executor();
+        if (executor.inEventLoop()) {
+            request0(n);
+        } else {
+            executor.execute(() -> request0(n));
+        }
+    }
+
+    private void request0(long n) {
+        if (obj3 == null) {
+            return;
+        }
+
+        if (n <= 0) {
+            onError(new IllegalArgumentException(
+                    "n: " + n + " (expected: > 0, see Reactive Streams specification rule 3.9)"));
+            return;
+        }
+
+        if (inOnNext) {
+            // re-entrance
+            if (n == 1) {
+                demand++;
+            } else {
+                demand += 2;
+            }
+            return;
+        }
+
+        if (n >= 3) {
+            inOnNext = true;
+            try {
+                if (obj1 != null) {
+                    final T item = obj1;
+                    obj1 = null;
+                    onNext(item);
+                }
+                if (obj2 != null) {
+                    final T item = obj2;
+                    obj2 = null;
+                    onNext(item);
+                }
+                if (obj3 != null) {
+                    final T item = obj3;
+                    obj3 = null;
+                    onNext(item);
+                }
+            } finally {
+                inOnNext = false;
+            }
+            onComplete();
+        } else {
+            demand += n;
+            if (obj1 != null) {
+                final T item = obj1;
+                obj1 = null;
+                inOnNext = true;
+                try {
+                    onNext(item);
+                } finally {
+                    inOnNext = false;
+                }
+                demand--;
+            }
+
+            if (obj2 != null && demand > 0) {
+                final T item = obj2;
+                obj2 = null;
+                try {
+                    onNext(item);
+                } finally {
+                    inOnNext = false;
+                }
+                demand--;
+            }
+
+            if (obj3 != null && demand > 0) {
+                final T item = obj3;
+                obj3 = null;
+                onNext(item);
```

Comment:
Don't we need to set `inOnNext` as `true` so that we can return it early in this method?

Reply:
`if (obj3 == null)` is checked at the beginning of the method.